moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h

```diff
@@ -756,36 +783,81 @@ class MoveGroupInterface
       This applies a number of hard-coded default grasps */
   MoveItErrorCode pick(const std::string& object);
 
+  /** \brief Pick up an object (with returning planning result)
+
+      This applies a number of hard-coded default grasps */
+  MoveItErrorCode pick(const std::string& object, PickPlan& plan);
+
   /** \brief Pick up an object given a grasp pose */
   MoveItErrorCode pick(const std::string& object, const moveit_msgs::Grasp& grasp);
 
+  /** \brief Pick up an object given a grasp pose  (with returning planning result) */
+  MoveItErrorCode pick(const std::string& object, const moveit_msgs::Grasp& grasp, PickPlan& plan);
+
   /** \brief Pick up an object given possible grasp poses
 
       if the vector is left empty this behaves like pick(const std::string &object) */
   MoveItErrorCode pick(const std::string& object, const std::vector<moveit_msgs::Grasp>& grasps);
 
+  /** \brief Pick up an object given possible grasp poses (with returning planning result)
+
+      if the vector is left empty this behaves like pick(const std::string &object) */
+  MoveItErrorCode pick(const std::string& object, const std::vector<moveit_msgs::Grasp>& grasps, PickPlan& plan);
+
   /** \brief Pick up an object
 
       calls the external moveit_msgs::GraspPlanning service "plan_grasps" to compute possible grasps */
-  MoveItErrorCode planGraspsAndPick(const std::string& object = "");
+  MoveItErrorCode planGraspsAndPick();
+
+  /** \brief Pick up an object (with returning planning result)
+
+      calls the external moveit_msgs::GraspPlanning service "plan_grasps" to compute possible grasps */
+  MoveItErrorCode planGraspsAndPick(PickPlan& plan);
+
+  /** \brief Pick up an object
+
+      calls the external moveit_msgs::GraspPlanning service "plan_grasps" to compute possible grasps */
+  MoveItErrorCode planGraspsAndPick(const std::string& object);
+
+  /** \brief Pick up an object (with returning planning result)
+
+      calls the external moveit_msgs::GraspPlanning service "plan_grasps" to compute possible grasps */
+  MoveItErrorCode planGraspsAndPick(const std::string& object, PickPlan& plan);
 
   /** \brief Pick up an object
 
       calls the external moveit_msgs::GraspPlanning service "plan_grasps" to compute possible grasps */
   MoveItErrorCode planGraspsAndPick(const moveit_msgs::CollisionObject& object);
 
+  /** \brief Pick up an object (with returning planning result)
+
+      calls the external moveit_msgs::GraspPlanning service "plan_grasps" to compute possible grasps */
+  MoveItErrorCode planGraspsAndPick(const moveit_msgs::CollisionObject& object, PickPlan& plan);
```

Comment:
Instead of duplicating all these functions can't you just specify a default value for PickPlan in the header?

Reply:
Since the `PickPlan` parameter need to be reference to store a plan, we can't directly put a default value for that. (a default value is a temporal object then the parameter should be const reference)

I you would like to use default parameters, solutions I came up with are,
1. Change the parameter type from `PickPlan&` to `PickPlan*`. so we can use `nullptr` as a default value. But this seems to be a little inconsistent with other MoveIt APIs.
2. Define a placeholder object statically and use it as a default value.

Reply:
IMO it should be `PickPlan&` to match the signature of `plan()`.
Alternative to using a dummy object I think all trivial overloads should call each other and be moved to the header (like getter()/setter() in RobotState).

Reply:
As @aochiai already wrote, it's not possible to pass a default argument.
I agree with @henningkayser to make all those functions call each other.